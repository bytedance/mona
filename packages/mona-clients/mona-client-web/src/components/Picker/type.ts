import React from 'react';
// import { PickerProps } from '@bytedance/mona';

export enum PickerCellMovingStatus {
  normal = 'normal',
  moving = 'moving',
  scrolling = 'scrolling',
}
export type ValueType = string | number;

export interface PickerData {
  value: ValueType;
  label: React.ReactNode;
  children?: PickerData[];
}

export interface PickerViewProps {
  data: PickerData[];


  cols?: number;
  rows?: number;
  value: ValueType[];
  onColumnChange?: (value: ValueType[], index: number) => void;
}
