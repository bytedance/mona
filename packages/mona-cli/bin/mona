#! /usr/bin/env node

const mona = require('../dist').default;
mona();
