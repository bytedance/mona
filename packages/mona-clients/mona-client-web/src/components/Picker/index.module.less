.monaPickerModal {
  z-index: 9999;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.8);
}

.pickerWrapper {
  font-size: 36px;
  background-color: white;

  .pickerView {
    position: relative;
    width: 100%;
    z-index: 2;
    margin: 0.5em 0;

    &Multi {
      color: #1d2129;
      font-size: 1em;
    }

    &Multi,
    &Selection {
      display: flex;
      align-items: center;
      position: absolute;
      left: 0;
      top: 0;
      width: 100%;
      height: 100%;
    }

    &Column {
      flex: 1;
      height: 100%;
      overflow: hidden;
      text-align: center;

      &ItemWrap {
        box-sizing: content-box;
        will-change: transform;
      }
    }

    &Selection {
      will-change: transform;
      flex-direction: column;
      pointer-events: none;

      &MaskTop {
        width: 100%;
        flex: 1;
        background: rgba(255, 255, 255, 0.404);
      }

      &MaskBottom {
        width: 100%;
        flex: 1;
        background: rgba(255, 255, 255, 0.404);
      }

      &Bar {
        width: 100%;
        height: 2em;
        flex: 0 0 auto;
        top: 1px solid #e5e6eb;
        bottom: 1px solid #e5e6eb;
        border: 1px solid #e5e6eb;
      }
    }
  }
  &Header {
    padding: 0 30px;
    height: 2.3em;
    background: white;
    display: flex;
    justify-content: space-between;
    align-items: center;
    color: #34c1f5;
    font-weight: normal;
    border: 1px solid #e5e6eb;

    &Cancel {
      font-size: 1em;
    }
    &Cofirm {
      font-size: 1em;
    }
  }
}
