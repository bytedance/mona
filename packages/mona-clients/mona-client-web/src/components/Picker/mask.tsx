import React from 'react';
import { formatTouchEvent } from '../utils';
import styles from './index.module.less';

const stopPropagation = (e: any) => {
  e.stopPropagation();
};
const maskStyle: React.CSSProperties = {
  bottom: 0,
  position: 'absolute',
  left: 0,
  right: 0,
  background: 'white',
};
interface IProps {
  visible: boolean;
  onConfirm: (e: any) => any;
  onCancel: (e: any) => any;
}
const PickerMask: React.FC<IProps> = ({ visible, children, onConfirm, onCancel }) => {
  if (visible) {
    return (
      <div
        className={styles.monaPickerModal}
        onTouchStart={e => {
          stopPropagation(e);
          onCancel?.(formatTouchEvent({ event: e, type: 'change' }));
        }}
        onClick={stopPropagation}
        onMouseDown={stopPropagation}
      >
        <div style={maskStyle} onTouchStart={stopPropagation}>
          <div className={styles.pickerWrapper}>
            <div className={styles.pickerWrapperHeader}>
              <div
                className={styles.pickerWrapperHeaderCancel}
                onTouchStart={(e: React.TouchEvent) => {
                  onCancel?.(formatTouchEvent({ event: e, type: 'change' }));
                }}
              >
                取消
              </div>
              <div
                className={styles.pickerWrapperHeaderConfirm}
                onTouchStart={e => {
                  onConfirm?.(formatTouchEvent({ event: e, type: 'change' }));
                }}
              >
                确定
              </div>
            </div>
            {children}
          </div>
        </div>
      </div>
    );
  }

  return null;
};

export default PickerMask;
