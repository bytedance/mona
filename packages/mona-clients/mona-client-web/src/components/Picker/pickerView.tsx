import React, { useRef, useState, useEffect, forwardRef, Ref, useImperativeHandle } from 'react';
import Cascader from './components/cascader';
import { PickerViewProps, ValueType, PickerData } from './type';
import styles from './index.module.less';
// 引入less失败
console.log('style', styles);

// disabled value
const prefixCls = 'mona';
const hideEmptyCols = false;

const PickerView = forwardRef((props: PickerViewProps, ref: Ref<{ getData: () => ValueType[] }>) => {
  const { cols = 5, rows = 5, data, disabled = false, value, onColumnChange } = props;
  const [scrollValue, setScrollValue] = useState(value);

  const [itemHeight, setItemHeight] = useState(0);
  const [wrapperHeight, setWrapperHeight] = useState(0);

  const wrapperRef = useRef<HTMLDivElement>(null);
  const barRef = useRef<HTMLDivElement | null>(null);
  const scrollValueRef = useRef(value);

  scrollValueRef.current = scrollValue;
  useImperativeHandle(ref, () => ({
    getData() {
      return scrollValueRef.current;
    },
  }));
  function _onColumnChange(value: ValueType[], index: number) {
    setScrollValue(value);
    onColumnChange?.(value, index);
  }

  useEffect(() => {
    if (wrapperRef.current) {
      setWrapperHeight(wrapperRef.current.offsetHeight);
    }
    if (barRef.current) {
      setItemHeight(barRef.current.clientHeight);
    }
  }, []);

  const newItemStyle = {
    height: `${itemHeight}px`,
  };

  return (
    <>
      <div
        className={`${styles.pickerView} all-border-box `}
        style={{ height: `${itemHeight * rows}px` }}
        ref={wrapperRef}
      >
        <Cascader
          prefixCls={prefixCls || ''}
          cols={cols}
          itemStyle={newItemStyle}
          data={data as PickerData[]}
          selectedValue={scrollValue}
          onColumnChange={_onColumnChange}
          itemHeight={itemHeight}
          wrapperHeight={wrapperHeight}
          disabled={disabled}
          rows={rows}
          hideEmptyCols={hideEmptyCols}
        />

        <div className={styles.pickerViewSelection}>
          <div className={styles.pickerViewSelectionMaskTop} />
          <div ref={barRef} className={styles.pickerViewSelectionBar} />
          <div className={styles.pickerViewSelectionMaskTop} />
        </div>
      </div>
    </>
  );
});

export default PickerView;
