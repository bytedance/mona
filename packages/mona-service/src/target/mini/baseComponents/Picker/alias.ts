import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { PickerAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  name: propAliasMap.name,
  mode: propAliasMap.mode,
  value: propAliasMap.value,
  disabled: propAliasMap.disabled,
  onCancel: propAliasMap.onCancel,
  range: propAliasMap.range,
  rangeKey: propAliasMap.rangeKey,
  start: propAliasMap.start,
  end: propAliasMap.end,
  fields: propAliasMap.fields,
  customItem: propAliasMap.customItem,
  onChange: propAliasMap.onChange,
  onColumnChange: propAliasMap.onColumnChange,
};

export default alias;
