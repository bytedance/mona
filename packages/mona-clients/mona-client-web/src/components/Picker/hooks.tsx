import { PickerProps, TouchEvent } from '@bytedance/mona';

import { PickerData } from './type';
import { genEvent } from '../utils';
import { useMemo, useRef, useState } from 'react';

export const useProps = (props: PickerProps) => {
  const { mode } = props;
  const propsRef = useRef<PickerProps>(props);
  propsRef.current = props;
  const finalData = useMemo(() => {
    let finalData: {
      data: PickerData[];
      value: any;
    } = {
      data: [],
      value: props.value,
    };
    switch (mode) {
      case 'selector':
        finalData.data = genData([props.range!], props.rangeKey);
        finalData.value = [props.value];
        break;
      case 'multiSelector':
        finalData.value = Object.assign(
          Array.from({ length: props.range?.length || 0 }, () => 0),
          props.value,
        );
        finalData.data = genData(props.range || [], props.rangeKey);

        break;
      case 'date':
        finalData = genDateData(props.value, props.start, props.end, props.fields);

        break;
      case 'time':
        finalData = genTimeData(props.value, props.start, props.end);
        break;
      case 'region':
        break;
      default:
        const exhaustiveCheck: never = props;
        //@ts-ignore
        finalData.value = exhaustiveCheck.value;
        break;
    }
    return {
      data: finalData?.data as unknown as PickerData[],
      value: finalData.value,
    };
    //@ts-ignore
  }, [mode, props.value, props.start, props.end, props.fields, props.range, props.rangeKey]);
  const handler = useMemo(() => {
    let handleConfirm, handleChange, handleCancel;
    handleConfirm = propsRef.current.onChange;
    handleCancel = propsRef.current.onCancel;

    switch (mode) {
      case 'selector':
        handleConfirm = (e: TouchEvent) => {
          e.detail.value = e.detail.value?.[0];
          propsRef.current.onChange?.(e);
        };
        break;
      case 'multiSelector':
        handleChange = (value: any[], idx: number) => {
          //@ts-ignore
          propsRef.current.onColumnChange?.(
            genEvent({
              type: 'change',
              detail: {
                column: +idx,
                value: value[idx],
              },
            }),
          );
        };
        break;
      case 'date':
        handleConfirm = (e: TouchEvent) => {
          const value = e.detail.value;
          const y = finalData.data[value[0]];
          const m = y?.children?.[value[1]];
          const d = m?.children?.[value[2]];

          e.detail.value = [y.label, m?.label, d?.label].filter(item => item !== undefined).join(':');
          propsRef.current?.onChange?.(e);
        };
        break;
      case 'time':
        handleConfirm = (e: TouchEvent) => {
          const value = e.detail.value;
          const h = finalData.data[value[0]];
          const m = h.children?.[value[1]];
          e.detail.value = [h.label, m?.label].join(':');
          propsRef.current?.onChange?.(e);
        };

        break;

      case 'region':
        break;

      default:
        break;
    }
    return {
      onColumnChange: handleChange,
      onConfirm: handleConfirm,
      onCancel: handleCancel,
    };
  }, [finalData.data]);
  return {
    ...handler,
    ...finalData,
  };
};

export function useRefState<T>(
  initialValue: T | (() => T),
): [T, React.MutableRefObject<T>, React.Dispatch<React.SetStateAction<T>>] {
  const [state, setState] = useState<T>(initialValue);
  const stateRef = useRef<T>(state);
  stateRef.current = state;
  return [state, stateRef, setState];
}

const genArr = (start: string, end: string) => {
  return Array.from({ length: +end + 1 - +start }, (_, idx) => {
    return idx + Number(start);
  });
};
const genTimeData = (value?: string, start: string = '00:00', end: string = '23:59') => {
  const [sH, sM] = start.split(':');
  const [eH, eM] = end.split(':');
  const AllMmArr = genArr('00', '59').map(m => {
    return m < 10 ? `0${m}` : `${m}`;
  });
  const mmArr = genArr(sM, eM).map(m => {
    return m < 10 ? `0${m}` : `${m}`;
  });
  const hhArr = genArr(sH, eH).map(m => {
    return m < 10 ? `0${m}` : `${m}`;
  });
  let timeArr = genData([hhArr, AllMmArr]);
  if (Number(eH) - Number(sH) >= 1) {
    timeArr = genData([hhArr, AllMmArr]);
    timeArr[0].children = genData([
      genArr(sM, 59 + '').map(m => {
        return +m < 10 ? `0${m}` : `${m}`;
      }),
    ]);

    timeArr[timeArr.length - 1].children = genData([
      genArr('0', eM).map(m => {
        return +m < 10 ? `0${m}` : `${m}`;
      }),
    ]);
  } else {
    timeArr = genData([hhArr, mmArr]);
  }
  const valueArr = value?.split(':') || [sH, sM];

  const hourIndex = hhArr.findIndex(item => +valueArr[0] === +item);
  const mIndx = timeArr
    .find(item => item.value === hourIndex)
    .children?.find((item: any) => +valueArr[1] === +item.label)?.value;
  return {
    value: [hourIndex, mIndx],
    data: timeArr,
  };
};

const genDateData = (value?: string, start: string = '2000-10-01', end: string = '2021-10-10', fields?: string) => {
  const [startYYYY, startMM, startDD] = start.split('-');
  const [endYYYY, endMM, endDD] = end.split('-');

  const YYYYArr = genArr(startYYYY, endYYYY);
  const MMArr = genArr(startMM, endMM);
  const isYear = fields === 'year' ? [] : undefined;
  const isMounth = fields === 'month' ? [] : undefined;

  let finalValue = genYearDataArr(YYYYArr, !isYear, !isMounth);
  let monthValue, dayValue, yearValue;
  const valueArr = value?.split('-') ?? [];
  if (!isYear) {
    if (+endYYYY - +startYYYY < 1) {
      finalValue[0].children = genMounthDataArr(MMArr, !isMounth);

      if (!isMounth) {
        if (+endMM - +startMM < 1) {
          finalValue[0].children[0].children = genData([dayArr[+startMM - 1].slice(+startDD - 1, +endDD)]);
        } else {
          finalValue[0].children[0].children = genData([dayArr[+startMM - 1].slice(+startDD - 1)]);
          finalValue[0].children[finalValue[0].children.length - 1].children = genData([
            dayArr[+endMM - 1].slice(0, +endDD),
          ]);
        }
      }
    } else {
      finalValue[0].children = genMounthDataArr(genArr(startMM, 12 + ''), !isMounth ? [] : undefined);
      finalValue[finalValue.length - 1].children = genMounthDataArr(genArr(1 + '', endMM), !isMounth);
      if (!isMounth) {
        finalValue[0].children[0].children = genData([dayArr[+startMM - 1].slice(+startDD - 1)]);
        finalValue[finalValue.length - 1].children[finalValue[finalValue.length - 1].children.length - 1].children =
          genData([dayArr[+endMM - 1].slice(0, +endDD)]);
      }
    }
  }
  if (isYear) {
    if (valueArr[0] !== undefined) {
      yearValue = finalValue.find(item => +item.label === +valueArr[0]);
    }
  } else if (isMounth) {
    if (valueArr[0] !== undefined) {
      yearValue = finalValue.find(item => +item.label === +valueArr[0]);
    }
    if (yearValue && valueArr[1] !== undefined) {
      monthValue = yearValue.children?.find(item => +item.label === +valueArr[1]);
    }
  } else {
    if (valueArr[0] !== undefined) {
      yearValue = finalValue.find(item => +item.label === +valueArr[0]);
    }
    if (yearValue && valueArr[1] !== undefined) {
      monthValue = yearValue.children?.find(item => +item.label === +valueArr[1]);
    }
    if (monthValue && valueArr[2] !== undefined) {
      dayValue = monthValue.children?.find(item => +item.label === +valueArr[2]);
    }
  }

  return {
    data: finalValue,
    value: [yearValue?.value, monthValue?.value, dayValue?.value].filter(item => item !== undefined),
  };
};

// TODO: 闰年
const dayArr = [31, 28, 31, 30, 31, 30, 31, 30, 31, 30, 31, 30].map(item => genArr(1 + '', item + ''));
const monthArr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

const genMounthDataArr = (MMarr: any[], children: any) => {
  return MMarr.map((item, idx) => {
    return {
      value: idx,
      label: item,
      children: children ? genData([dayArr[+item - 1]]) : undefined!,
    };
  });
};

const genYearDataArr = (YYYYArr: any[], yearChildren: any, monthChildren: any) => {
  const MMarr = monthArr.map((item, idx) => {
    return {
      value: idx,
      label: item,
      children: monthChildren ? genData([dayArr[+item - 1]]) : undefined,
    };
  });
  return YYYYArr.map((item, idx) => {
    return {
      value: idx,
      label: item,
      children: yearChildren ? MMarr : undefined!,
    };
  });
};

const genData = (data: (string | number | boolean)[][], rangeKey?: string) => {
  let children: any[];
  data
    .slice()
    .reverse()
    .forEach(arr => {
      children = arr.map((value, idx) => {
        if (rangeKey && value !== null && typeof value === 'object') {
          return {
            value: idx,
            label: value[rangeKey],
            children,
          };
        }
        return {
          value: idx,
          label: value,
          children,
        };
      });
    });
  //@ts-ignore
  return children;
};
