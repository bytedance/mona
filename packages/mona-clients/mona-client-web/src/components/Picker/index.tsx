import { PickerProps } from '@bytedance/mona';
import React, { useCallback, useRef, useState } from 'react';
import { useProps } from './hooks';
import PickerMask from './mask';
import PickerView from './pickerView';
import { TouchEvent } from '@bytedance/mona';

const Picker: React.FC<PickerProps> = props => {
  const [visible, setVisible] = useState(false);
  const newProps = useProps(props);
  const pickerRef = useRef(newProps.value);
  const propsRef = useRef(newProps);

  const handleCancel = useCallback((e: TouchEvent) => {
    setVisible(false);
    propsRef.current?.onCancel?.({
      ...e,
      detail: { value: pickerRef.current.getData() },
    });
  }, []);
  const handleConfirm = useCallback((e: TouchEvent) => {
    setVisible(false);
    propsRef.current?.onConfirm?.({
      ...e,
      detail: { value: pickerRef.current.getData() },
    });
  }, []);
  return (
    <div>
      <div onTouchStart={() => !props.disabled && setVisible(true)}>
        {props.children}
        <PickerMask visible={visible} onCancel={handleCancel} onConfirm={handleConfirm}>
          <PickerView ref={pickerRef} {...newProps}></PickerView>
        </PickerMask>
      </div>
    </div>
  );
};
export default Picker;
