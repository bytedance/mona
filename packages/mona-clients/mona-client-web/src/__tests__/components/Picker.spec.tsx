import React from 'react';
import { mount } from 'enzyme';
import Picker from '../../components/Picker';
import mountTest from '../../../../../tests/shared/mountTest';
// import handlerTest from '../../../../../tests/shared/handlerTest';
import { PickerProps } from '@bytedance/mona';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure } from 'enzyme';

configure({ adapter: new Adapter() });

describe('web component: Picker', () => {
  mountTest(Picker)
  // handlerTest({ Component: Picker, firstTag: 'div' });

  it('should render correctly', () => {
    expect(mount(<Picker mode="region" />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: PickerProps = {
      mode: 'region',
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<Picker {...props}>hello</Picker>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
