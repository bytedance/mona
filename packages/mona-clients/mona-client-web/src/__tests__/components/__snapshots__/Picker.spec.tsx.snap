// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: Picker should have right properties 1`] = `
LoadedCheerio {
  "0": <div>
    <div>
      hello
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: Picker should render correctly 1`] = `
LoadedCheerio {
  "0": <div>
    <div />
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
